packages/react/src/Textarea/Textarea.tsx

```diff
@@ -68,37 +77,128 @@ const Textarea = React.forwardRef<HTMLTextAreaElement, TextareaProps>(
+
+        return () => {
+          characterCounterRef.current?.cleanup()
+        }
+      }
+    }, [characterLimit, value, defaultValue])
+
+    // Update character count when value changes
+    useEffect(() => {
+      if (characterLimit && characterCounterRef.current) {
+        const currentValue = value !== undefined ? String(value) : ''
+        characterCounterRef.current.updateCharacterCount(currentValue.length, characterLimit)
+      }
+    }, [value, characterLimit])
+
+    // Handle textarea change with character counter
+    const handleTextareaChange = useCallback(
+      (e: React.ChangeEvent<HTMLTextAreaElement>) => {
+        if (characterLimit && characterCounterRef.current) {
+          characterCounterRef.current.updateCharacterCount(e.target.value.length, characterLimit)
+        }
+        onChange?.(e)
+      },
+      [onChange, characterLimit],
+    )
+
+    const isValid = isOverLimit ? 'error' : validationStatus
+
     return (
-      <TextInputBaseWrapper
-        validationStatus={validationStatus}
-        disabled={disabled}
-        block={block}
-        contrast={contrast}
-        className={className}
-      >
-        <textarea
-          value={value}
-          data-resize={resize}
-          aria-required={required}
-          aria-invalid={validationStatus === 'error' ? 'true' : 'false'}
-          ref={ref}
+      <>
+        <TextInputBaseWrapper
+          validationStatus={isValid}
           disabled={disabled}
-          rows={rows}
-          cols={cols}
-          className={classes.TextArea}
-          style={{
-            minHeight,
-            maxHeight,
-            ...style,
-          }}
-          {...rest}
-        />
-      </TextInputBaseWrapper>
+          block={block}
+          contrast={contrast}
+          className={className}
+        >
+          <textarea
+            value={value}
+            defaultValue={defaultValue}
+            data-resize={resize}
+            aria-required={required}
+            aria-invalid={isValid === 'error' ? 'true' : 'false'}
+            ref={ref}
+            disabled={disabled}
+            rows={rows}
+            cols={cols}
+            className={classes.TextArea}
+            onChange={handleTextareaChange}
+            style={{
+              minHeight,
+              maxHeight,
+              ...style,
+            }}
+            {...rest}
+            aria-describedby={
+              characterLimit
+                ? [characterCountStaticMessageId, rest['aria-describedby']].filter(Boolean).join(' ') || undefined
+                : rest['aria-describedby']
+            }
+          />
+        </TextInputBaseWrapper>
+        {characterLimit && (
+          <>
+            <VisuallyHidden id={characterCountLiveRegionId} aria-live="polite" role="status">
+              {screenReaderMessage}
+            </VisuallyHidden>
+            <VisuallyHidden id={characterCountStaticMessageId}>
```

Comment:
This is hidden as we want to include it for screen reader users, but including the actual character counter span that updates when a user types was causing double announcements.